### PR TITLE
Rewrite `impl_runtime_apis!` and `decl_runtime_apis!` as proc-macro

### PR DESCRIPTION
**Pull request by bkchr**
This is a complete rewrite of both macros as `proc-macros`. 
Using `proc-macro` brings us certain advantages, like better error reporting and a more readable implementation.

### CODE REVIEW

File: core/sr-api-macros/src/impl_runtime_apis.rs [hunk 1]
```diff
@@ -0,0 +1,530 @@
...
+		Err(Error::new(span, "Missing `Block` generic parameter."))
+	} else {
+		match &generics.arguments {
+			PathArguments::AngleBracketed(ref args) => {
+				args.args.first().and_then(|v| match v.value() {
+					GenericArgument::Type(Type::Path(block)) => Some(block),
+					_ => None
+				}).ok_or_else(|| Error::new(args.span(), "Missing `Block` generic parameter."))
+			},
+			PathArguments::None => unreachable!(),
```
Comment (gnunicorn) [inline core/sr-api-macros/src/impl_runtime_apis.rs]: why not have the code within the `if ...is_emtpy()` here instead and remove one indentation level? Is there a difference between `.is_empty()` and `==PathArguments::None`?

File: core/sr-api-macros/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+/// # use primitives::hash::H256;
+/// # use runtime_primitives::traits::{BlakeTwo256, GetNodeBlockType, Extrinsic as ExtrinsicT};
+/// #
+/// # // All the stuff we need to declare our `Block`
+/// # pub type BlockNumber = u64;
+/// # pub type DigestItem = runtime_primitives::generic::DigestItem<H256, u64>;
+/// # pub type Digest = runtime_primitives::generic::Digest<DigestItem>;
+/// # #[derive(Clone, PartialEq, Eq, Encode, Decode, Debug)]
+/// # pub struct Extrinsic {}
+/// # impl serde::Serialize for Extrinsic
```
Comment (gnunicorn) [inline core/sr-api-macros/src/lib.rs]:
```suggestion
/// #
/// # impl serde::Serialize for Extrinsic {
```

File: core/sr-api-macros/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+/// # {
```
Comment (gnunicorn) [inline core/sr-api-macros/src/lib.rs]:
```suggestion
```

File: core/sr-api-macros/src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+/// Fold the given `FnDecl` to make it usable on the client side.
+pub fn fold_fn_decl_for_client_side(
+	mut input: FnDecl,
+	block_id: &TokenStream,
+	crate_: &TokenStream
+) -> FnDecl {
+	// Add `&` to all parameter types.
+	// TODO: Do not enforce `&`.
```
Comment (gnunicorn) [inline core/sr-api-macros/src/utils.rs]: No commit of `//TODO` in master unless there is a link to the issue for it, please.

File: core/sr-api-macros/tests/decl_and_impl.rs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+use primitives::AuthorityId;
+use substrate_client::error::Result;
+
+// All the stuff we need to declare our `Block`
+pub type BlockNumber = u64;
+pub type DigestItem = runtime_primitives::generic::DigestItem<H256, u64>;
+pub type Digest = runtime_primitives::generic::Digest<DigestItem>;
+#[derive(Clone, PartialEq, Eq, Encode, Decode, Debug)]
+pub struct Extrinsic {}
+impl serde::Serialize for Extrinsic
```
Comment (gnunicorn) [inline core/sr-api-macros/tests/decl_and_impl.rs]:
```suggestion

impl serde::Serialize for Extrinsic {
```

File: core/sr-api-macros/tests/decl_and_impl.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+{
```
Comment (gnunicorn) [inline core/sr-api-macros/tests/decl_and_impl.rs]:
```suggestion
```

File: core/sr-api-macros/tests/decl_and_impl.rs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+		}
+	}
+}
+
+#[test]
+fn test_client_side_function_signature() {
+	let _test: fn(&RuntimeApi, &BlockId<Block>, &u64) -> Result<()>  = RuntimeApi::test;
+	let _something_with_block: fn(&RuntimeApi, &BlockId<Block>, &Block) -> Result<Block> =
+		RuntimeApi::something_with_block;
+}
```
Comment (gnunicorn) [inline core/sr-api-macros/tests/decl_and_impl.rs]: Is it possible to write a test that shows that the traits have to be met? Like one that implements a method with other signature than the trait it is supposed to implement? That'd be a great safe guard for people to have insurance about.

File: core/sr-api-macros/src/utils.rs (hunk 4 above)
Comment (bkchr) [reply core/sr-api-macros/src/utils.rs]: Ahh! I oversaw this one^^

File: core/sr-api-macros/tests/decl_and_impl.rs (hunk 7 above)
Comment (bkchr) [reply core/sr-api-macros/tests/decl_and_impl.rs]: Added a test to ensure that: https://github.com/paritytech/substrate/pull/1174/files#diff-462516f88a68cb02cf39ce86f563af73R351

File: node/runtime/src/lib.rs [hunk 8]
```diff
@@ -452,26 +287,41 @@ impl_runtime_apis! {
 			Executive::finalise_block()
 		}
 
-		fn inherent_extrinsics(data: InherentData) -> Vec<UncheckedExtrinsic> {
-			data.create_inherent_extrinsics()
+		fn inherent_extrinsics(data: runtime_primitives::InherentData) -> Vec<<Block as BlockT>::Extrinsic> {
+			let mut inherent = Vec::new();
+
+			inherent.extend(
```
Comment (gavofyork) [inline node/runtime/src/lib.rs]: manual amalgamation code?

Comment (bkchr) [reply node/runtime/src/lib.rs]: Yeah, I reverted in parts my macro here. I ran into some problems. As I will need to revisit the `InherentData` stuff anyway, I did not wanted to waste too much time on this topic in this pr.

**Pull request merged**